File: Data Structures & Algorithms/maximum-subarray/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    maxSubArray(nums) {
        /**
         * SubProblem : s[i:] max sum
         * relation :
         *    currSum = nums[i]+currSum;
         *    max = math.max(max,currSum);
         */

        let currSum = 0;
        let max = -Infinity;

        for(let i = nums.length-1 ; i>-1;i--){
            currSum = Math.max(currSum+nums[i],nums[i]);
            max = Math.max(max,currSum);
        }
        return max;
    }
}

File: Data Structures & Algorithms/target-sum/submission-1.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        n = len(nums)
        result = 0
        dp = {}

        def dfs(i,sum):
            if i == n:
                return 1 if sum == target else 0
            
            if (i,sum) in dp:
                return dp[(i,sum)]
            
            dp[(i,sum)] = (dfs(i+1,sum+nums[i]) + dfs(i+1,sum-nums[i]))

            return dp[(i,sum)]
    
        return dfs(0,0)
        